exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,384 @@
+}
+
+// Export exports the provide metric record to prometheus.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var forEachError error
+	checkpointSet.ForEach(func(record export.Record) {
+		agg := record.Aggregator()
+
+		desc := record.Descriptor()
+
+		labels := record.Labels()
+		if labels.Encoder() != e {
+			// TODO: This case could be handled by directly
+			// encoding the labels at this point, but presently it
+			// should not occur.
+			//
+			// copy from datadog exporter.
+			panic("Should have self-encoded labels")
+		}
+		mKey := metricKey{
+			desc: desc,
+			// TODO: check if encoder is the exporter.
+			encoded: record.Labels().Encoded(),
+		}
+
+		// TODO(paivagustavo): how to choose between Histogram and Summary?
+		if all, ok := agg.(aggregator.AllValues); ok {
+			values, err := all.AllValues()
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			o, err := e.getHistogram(record, mKey)
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			desc := record.Descriptor()
+			for _, v := range values {
+				o.Observe(v.CoerceToFloat64(desc.NumberKind()))
+			}
+
+			return
+		}
+
+		if sum, ok := agg.(aggregator.Sum); ok {
+			value, err := sum.Sum()
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			c, err := e.getCounter(record, mKey)
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			desc := record.Descriptor()
+			c.Add(value.CoerceToFloat64(desc.NumberKind()))
+
+			return
+		}
+
+		if gauge, ok := agg.(aggregator.LastValue); ok {
+			lv, _, err := gauge.LastValue()
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			g, err := e.getGauge(record, mKey)
+			if err != nil {
+				// TODO: handle this better when we have a more
+				//  sophisticated error handler mechanism for this ForEach method.
+				forEachError = err
+				return
+			}
+
+			desc := record.Descriptor()
+			g.Set(lv.CoerceToFloat64(desc.NumberKind()))
+
+			return
+		}
+	})
+
+	return forEachError
+}
+
+func (e *Exporter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	e.handler.ServeHTTP(w, r)
+}
+
+func (e *Exporter) Encode(kvs []core.KeyValue) string {
+	return e.labelEncoder.Encode(kvs)
+}
+
+func (e *Exporter) getCounter(record export.Record, mKey metricKey) (prometheus.Counter, error) {
+	e.Lock()
+	defer e.Unlock()
```

Comment:
technically can have a lock per metric type, but probably not worth it.

Reply:
We actually don't need mutexes here. Just removing it.